sdk/core/core/src/az_span.c

```diff
@@ -525,31 +532,39 @@ AZ_NODISCARD az_result az_span_append_i64toa(az_span destination, int64_t source
 static AZ_NODISCARD az_result
 _az_span_builder_append_u32toa(az_span self, uint32_t n, az_span* out_span)
 {
+  AZ_RETURN_IF_SPAN_CAPACITY_TOO_SMALL(self, 1);
+
   if (n == 0)
   {
-    return az_span_append_uint8(self, '0', out_span);
+    *out_span = az_span_append_uint8(self, '0');
+    return AZ_OK;
   }
 
   uint32_t div = 1000000000;
   uint32_t nn = n;
+  int32_t digit_count = 10;
   while (nn / div == 0)
   {
     div /= 10;
+    digit_count--;
   }
 
+  AZ_RETURN_IF_SPAN_CAPACITY_TOO_SMALL(self, digit_count);
```

Comment:
What about refactoring all this kind of methods to calculate the total size they will need, then validate there is enough capacity and then append?
Something does not look right if we are appending stuff and then returning not enough capacity.
If this is too much trouble, we might want to at least create a new issue to address this in the future and make sure methods don't call this MACRO multiple times and always calculate the size, validate and then append/copy/replace

Reply:
I don't know if that works well for composability across the various call sites and might make the calling code harder to understand. The required length can't always be calculated up front.

At the very least, I am going to defer taking any action on this feedback for now, because it isn't clear that this is a net benefit. Without both capacity and length (in the future), we might remove the macro because the check becomes simpler.

> Something does not look right if we are appending stuff and then returning not enough capacity.

I don't understand what you mean by that. Can you explain a bit more. We currently return enough capacity if what we are trying to append won't fit.

> make sure methods don't call this MACRO multiple times

Right now, the macro can be called multiple times. I don't know if it is a good idea to put so much extra logic in a marco that users have to know not to call it twice.